The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9980754909/job/27582893623) [(plain)](https://github.com/rust-lang/rust/commit/f9b9b604c55ec1f4212f68102690e96c20bb5b9c/checks/27582893623/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:692973e3d937129bcbf40652eb9f2f61becf3332)
Download action repository 'actions/upload-artifact@v4' (SHA:0b2256b8c012f0828dc542b3febcab082c67f72b)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#12 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#12 DONE 0.0s

#13 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#13 0.448   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#13 0.465 Collecting boolean-py==4.0
#13 0.472   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#13 0.490 Collecting chardet==5.1.0
---
#13 4.737 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 5.267 Collecting virtualenv
#13 5.317   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 5.515      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 29.0 MB/s eta 0:00:00
#13 5.559 Collecting distlib<1,>=0.3.7
#13 5.565   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 5.575      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 57.8 MB/s eta 0:00:00
#13 5.606 Collecting platformdirs<5,>=3.9.1
#13 5.612   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 5.648 Collecting filelock<4,>=3.12.2
#13 5.655   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 5.743 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.928 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 6.0s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      192448 kB
DirectMap2M:     8196096 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/f00f85091904530ee8f5b29251c53571e0062a76/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-f00f85091904530ee8f5b29251c53571e0062a76-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished `release` profile [optimized] target(s) in 28.64s
##[endgroup]
fmt check
##[error]Diff in /checkout/library/std/src/os/windows/fs.rs at line 472:
     #[unstable(feature = "windows_by_handle", issue = "63010")]
     fn file_index(&self) -> Option<u64>;
-
-    /// Returns the change time, which is the last time file metadata was changed, such as 
+    /// Returns the change time, which is the last time file metadata was changed, such as
     /// renames, attributes, etc.
     /// renames, attributes, etc.
-    /// 
+    ///
     /// This will return `None` if the `Metadata` instance was not created using the `FILE_BASIC_INFO` type.
     #[unstable(feature = "windows_change_time", issue = "121478")]
     fn change_time(&self) -> Option<u64>;
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_mir_transform/src/validate.rs" "/checkout/compiler/rustc_mir_transform/src/dump_mir.rs" "/checkout/compiler/rustc_mir_transform/src/shim/async_destructor_ctor.rs" "/checkout/compiler/rustc_mir_transform/src/errors.rs" "/checkout/compiler/rustc_mir_transform/src/remove_zsts.rs" "/checkout/compiler/rustc_mir_transform/src/jump_threading.rs" "/checkout/compiler/rustc_mir_transform/src/lower_intrinsics.rs" "/checkout/compiler/rustc_mir_transform/src/coroutine.rs" "/checkout/compiler/rustc_mir_transform/src/add_call_guards.rs" "/checkout/compiler/rustc_mir_transform/src/single_use_consts.rs" "/checkout/compiler/rustc_mir_transform/src/copy_prop.rs" "/checkout/compiler/rustc_mir_transform/src/ssa.rs" "/checkout/compiler/rustc_macros/src/extension.rs" "/checkout/compiler/rustc_macros/src/symbols/tests.rs" "/checkout/compiler/rustc_macros/src/hash_stable.rs" "/checkout/compiler/rustc_macros/src/lift.rs" "/checkout/compiler/rustc_macros/src/current_version.rs" "/checkout/compiler/rustc_macros/src/serialize.rs" "/checkout/compiler/rustc_macros/src/lib.rs" "/checkout/compiler/rustc_macros/src/type_foldable.rs" "/checkout/compiler/rustc_macros/src/symbols.rs" "/checkout/compiler/rustc_macros/src/query.rs" "/checkout/compiler/rustc_macros/src/type_visitable.rs" "/checkout/compiler/rustc_macros/src/diagnostics/subdiagnostic.rs" "/checkout/compiler/rustc_macros/src/diagnostics/error.rs" "/checkout/compiler/rustc_macros/src/diagnostics/diagnostic.rs" "/checkout/compiler/rustc_macros/src/diagnostics/diagnostic_builder.rs" "/checkout/compiler/rustc_macros/src/diagnostics/utils.rs" "/checkout/compiler/rustc_macros/src/diagnostics/mod.rs" "/checkout/compiler/rustc_macros/build.rs" "/checkout/library/std/src/os/xous/services/log.rs" "/checkout/library/std/src/os/xous/mod.rs" "/checkout/library/std/src/os/vita/fs.rs" "/checkout/library/std/src/os/vita/raw.rs" "/checkout/library/std/src/os/vita/mod.rs" "/checkout/library/std/src/os/solaris/fs.rs" "/checkout/library/std/src/os/solaris/raw.rs" "/checkout/library/std/src/os/solaris/mod.rs" "/checkout/library/std/src/os/windows/process.rs" "/checkout/library/std/src/os/windows/ffi.rs" "/checkout/library/std/src/os/windows/fs.rs" "/checkout/library/std/src/os/windows/raw.rs" "/checkout/library/std/src/os/windows/thread.rs" "/checkout/library/std/src/os/windows/io/handle.rs" "/checkout/library/std/src/os/windows/io/socket.rs" "/checkout/library/std/src/os/windows/io/raw.rs" "/checkout/library/std/src/os/windows/io/tests.rs" "/checkout/library/std/src/os/windows/io/mod.rs" "/checkout/library/std/src/os/windows/mod.rs" "/checkout/library/std/src/os/fd/net.rs" "/checkout/library/std/src/os/fd/raw.rs" "/checkout/library/std/src/os/fd/tests.rs" "/checkout/library/std/src/os/fd/owned.rs" "/checkout/library/std/src/os/fd/mod.rs" "/checkout/library/std/src/os/openbsd/fs.rs" "/checkout/library/std/src/os/openbsd/raw.rs" "/checkout/library/std/src/os/openbsd/mod.rs" "/checkout/library/std/src/os/horizon/fs.rs" "/checkout/library/std/src/os/horizon/raw.rs" "/checkout/library/std/src/os/horizon/mod.rs" "/checkout/library/std/src/os/ios/mod.rs" "/checkout/library/std/src/os/darwin/fs.rs" "/checkout/library/std/src/os/darwin/raw.rs" "/checkout/compiler/rustc_mir_transform/src/coroutine/by_move_body.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Wed Jul 17 20:12:08 UTC 2024
  network time: Wed, 17 Jul 2024 20:12:08 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/library/std/src/os/windows/fs.rs
+++ b/library/std/src/os/windows/fs.rs
@@ -471,6 +471,13 @@ pub trait MetadataExt {
     /// `fs::metadata` or `File::metadata`, then this will return `Some`.
     #[unstable(feature = "windows_by_handle", issue = "63010")]
     fn file_index(&self) -> Option<u64>;
+
+    /// Returns the change time, which is the last time file metadata was changed, such as
+    /// renames, attributes, etc
+    ///
+    /// This will return `None` if the `Metadata` instance was not created using the `FILE_BASIC_INFO` type.
+    #[unstable(feature = "windows_change_time", issue = "121478")]
+    fn change_time(&self) -> Option<u64>;
 }
 
 #[stable(feature = "metadata_ext", since = "1.1.0")]
@@ -498,6 +505,9 @@ impl MetadataExt for Metadata {
     }
     fn file_index(&self) -> Option<u64> {
         self.as_inner().file_index()
+    }
+    fn change_time(&self) -> Option<u64> {
+        self.as_inner().changed_u64()
     }
 }
 

--- a/library/std/src/sys/pal/windows/fs.rs
+++ b/library/std/src/sys/pal/windows/fs.rs
@@ -32,6 +32,7 @@ pub struct FileAttr {
     creation_time: c::FILETIME,
     last_access_time: c::FILETIME,
     last_write_time: c::FILETIME,
+    change_time: Option<c::FILETIME>,
     file_size: u64,
     reparse_tag: u32,
     volume_serial_number: Option<u32>,
@@ -377,6 +378,7 @@ impl File {
                 creation_time: info.ftCreationTime,
                 last_access_time: info.ftLastAccessTime,
                 last_write_time: info.ftLastWriteTime,
+                change_time: None, // Only available in FILE_BASIC_INFO
                 file_size: (info.nFileSizeLow as u64) | ((info.nFileSizeHigh as u64) << 32),
                 reparse_tag,
                 volume_serial_number: Some(info.dwVolumeSerialNumber),
@@ -413,6 +415,10 @@ impl File {
                     dwLowDateTime: info.LastWriteTime as u32,
                     dwHighDateTime: (info.LastWriteTime >> 32) as u32,
                 },
+                change_time: Some(c::FILETIME {
+                    dhLowDateTime: info.ChangeTime as c::DWORD,
+                    dhHighDateTime: (info.ChangeTime >> 32) as c::DWORD,
+                }),
                 file_size: 0,
                 reparse_tag: 0,
                 volume_serial_number: None,
@@ -957,6 +963,10 @@ impl FileAttr {
         to_u64(&self.creation_time)
     }
 
+    pub fn changed_u64(&self) -> Option<u64> {
+        self.change_time.as_ref().map(|c| to_u64(c))
+    }
+
     pub fn volume_serial_number(&self) -> Option<u32> {
         self.volume_serial_number
     }
@@ -976,6 +986,7 @@ impl From<c::WIN32_FIND_DATAW> for FileAttr {
             creation_time: wfd.ftCreationTime,
             last_access_time: wfd.ftLastAccessTime,
             last_write_time: wfd.ftLastWriteTime,
+            change_time: None,
             file_size: ((wfd.nFileSizeHigh as u64) << 32) | (wfd.nFileSizeLow as u64),
             reparse_tag: if wfd.dwFileAttributes & c::FILE_ATTRIBUTE_REPARSE_POINT != 0 {
                 // reserved unless this is a reparse point

